backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -428,6 +433,120 @@ public FutureExperimentRunDAO(
+      }
+    }
+
+    // If linked dataset version ids are not exists in dataset then no need to check dataset
+    // authorization.
+    if (linkedDatasetVersionIds.isEmpty()) {
+      return InternalFuture.completedInternalFuture(datasets);
+    }
+    return jdbi.withHandle(
+            handle -> {
+              // We have dataset version as a commit and dataset as a repository in MDB so added
+              // commit_hash filter here
+              return handle
+                  .createQuery(
+                      " SELECT commit_hash, repository_id FROM repository_commit WHERE commit_hash IN (<linkedDatasetVersionIds>) ")
+                  .bindList("linkedDatasetVersionIds", linkedDatasetVersionIds)
+                  .map(
+                      (rs, ctx) ->
+                          new AbstractMap.SimpleEntry<>(
+                              rs.getString("commit_hash"), rs.getLong("repository_id")))
+                  .list();
+            })
+        .thenCompose(
+            datasetVersionDatasetList -> {
+              // Convert list of dataset and dataset version mapping into Map
+              Map<String, String> datasetVersionDatasetMap = new HashMap<>();
+              for (var entry : datasetVersionDatasetList) {
+                datasetVersionDatasetMap.put(entry.getKey(), String.valueOf(entry.getValue()));
+              }
+              return InternalFuture.completedInternalFuture(datasetVersionDatasetMap);
+            },
+            executor)
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              // Validate linked dataset in dataset version exists in database records if not then
+              // we will throw the error
+              Set<String> datasetIds = new HashSet<>();
+              for (Artifact dataset : datasets) {
+                String datasetVersionId = dataset.getLinkedArtifactId();
+                if (datasetVersionDatasetMap.containsKey(datasetVersionId)
+                    && datasetVersionDatasetMap.get(datasetVersionId) != null) {
+                  datasetIds.add(datasetVersionDatasetMap.get(datasetVersionId));
+                } else {
+                  return InternalFuture.failedStage(
+                      new InvalidArgumentException(
+                          "Dataset not found for dataset version: " + datasetVersionId));
+                }
+              }
+
+              // If all dataset with linked dataset versions found in database then checking
+              // authorization for the current user for those datasets
+              List<InternalFuture<String>> internalFutures = new LinkedList<>();
+              for (String datasetId : datasetIds) {
+                try {
+                  internalFutures.add(
+                      InternalFuture.completedInternalFuture(datasetId)
+                          .thenCompose(
+                              id ->
+                                  checkEntityPermissionBasedOnResourceTypes(
```

Comment:
Note that you changed the `check` to return a value vs raise an exception in a different PR. You'll end up with the wrong behavior here.

Reply:
Ok, i will be managed it when different PR code will merge in this PR.